lib/dma.c

```diff
@@ -599,11 +600,32 @@ dma_controller_dirty_page_get(dma_controller_t *dma, vfu_dma_addr_t addr,
         return ERROR_INT(EINVAL);
     }
 
-    memcpy(bitmap, region->dirty_bitmap, size);
+    for (i = 0; i < (size_t)bitmap_size; i++) {
+        uint8_t val = region->dirty_bitmap[i];
```

Comment:
I don't think this is safe. What if we read a part of the bitmap some bits set, say:

00011111

Then another thread calls mark dirty on the some more bits from the byte, so that it becomes:

01111111

Then we do our atomic exchange with zero. We have val as 00011111 and have zeroed out the byte but pages 2 and 3 are still dirty.

Sadly I don't see how we can avoid the atomic exchange.

Reply:
We don't use `val` for the atomic exchange, we use `region->dirty_bitmap[i]`, see L621.

Reply:
Apologies - bad read on my part. This looks good.